src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -0,0 +1,263 @@
+
+  def lookupFun(ec: EvalContext)(name: String, typs: Seq[BaseType])(args: Seq[AST]): Option[() => Any] = {
+    require(typs.length == args.length)
+
+    val funTyp = TypeTag.funTypeFromArgs(typs)
+
+    lookup(name, funTyp).map { fn => (fn, typs) match {
+      case (f: UnaryFun[t, u], Seq(x)) =>
+        AST.evalCompose(ec, args(0))(f)
+      case (f: OptionUnaryFun[t, u], Seq(x)) =>
+        AST.evalFlatCompose(ec, args(0))(f)
+      case (f: BinaryFun[t, u, v], Seq(x, y)) =>
+        AST.evalCompose(ec, args(0), args(1))(f)
+      case (f: Arity3Fun[t, u, v, w], Seq(x, y, z)) =>
+        AST.evalCompose(ec, args(0), args(1), args(2))(f)
+      case (f: Arity4Fun[t, u, v, w, x], Seq(x, y, z, a)) =>
+        AST.evalCompose(ec, args(0), args(1), args(2), args(3))(f)
+      case _ =>
+        throw new RuntimeException(s"Internal hail error, bad binding in function registry for `$name' with argument types $typs: $fn")
+    }
+    }
+  }
+
+  def lookupFunReturnType(name: String, typs: Seq[BaseType]): Option[Type] =
+    lookup(name, TypeTag.funTypeFromArgs(typs)).map(_.retType)
+
+  def registerUnaryMethod[T, U](name: String, impl: T => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(hrt.typ), UnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerUnaryFun[T, U](name: String, impl: T => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryFunType(hrt.typ), UnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerOptionUnaryMethod[T, U](name: String, impl: T => Option[U])
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(hrt.typ), OptionUnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerOptionUnaryFun[T, U](name: String, impl: T => Option[U])
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryFunType(hrt.typ), OptionUnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerUnaryNAFilteredCollectionMethod[T, U](name: String, impl: TraversableOnce[T] => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(TArray(hrt.typ)), UnaryFun[IndexedSeq[_], U](hru.typ, { (ts: IndexedSeq[_]) =>
+      impl(ts.filter(t => t != null).map(_.asInstanceOf[T]))
+    }))
+    bind(name, UnaryMethodType(TSet(hrt.typ)), UnaryFun[Set[_], U](hru.typ, { (ts: Set[_]) =>
+      impl(ts.filter(t => t != null).map(_.asInstanceOf[T]))
+    }))
+  }
+
+  def registerBinary[T, U, V](name: String, impl: (T, U) => V)
+    (implicit hrt: HailRep[T], hru: HailRep[U], hrv: HailRep[V]) = {
+    bind(name, BinaryFunType(hrt.typ, hru.typ), BinaryFun[T, U, V](hrv.typ, impl))
+  }
+
+  def register3Ary[T, U, V, W](name: String, impl: (T, U, V) => W)
+    (implicit hrt: HailRep[T], hru: HailRep[U], hrv: HailRep[V], hrw: HailRep[W]) = {
+    bind(name, Arity3FunType(hrt.typ, hru.typ, hrv.typ), Arity3Fun[T, U, V, W](hrw.typ, impl))
+  }
+
+  def register4Ary[T, U, V, W, X](name: String, impl: (T, U, V, W) => X)
```

Comment:
Is it cleaner to have `registerBinary`, `register3Aray`, etc. as overloaded `register`?